### PR TITLE
setup training_speech importer (see #1576)

### PR DESCRIPTION
**Pull request by nicolaspanel**
importer for https://github.com/mozilla/DeepSpeech/issues/1576

Produces 3 `<target_dir>/ts_<ARCHIVE_NAME>_(dev|train|test).csv`  files

containing respectively

```
$ wc -l <target_dir>/ts_2018-09-28_fr*          
   5603 <target_dir>/ts_2018-09-28_fr_FR_dev.csv
   5603 <target_dir>/ts_2018-09-28_fr_FR_test.csv
  44811 <target_dir>/ts_2018-09-28_fr_FR_train.csv
  56017 total

```

and look like:
```
$ head -n 5 <target_dir>/training_speech_2018-09-28_fr_FR_*  
==> <target_dir>/ts_2018-09-28_fr_FR_dev.csv <==
wav_filename,wav_filesize,transcript
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0004.wav,219564,aussitôt comme d'habitude encore la plate forme du fort saint jean s'était couverte de curieux
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0019.wav,161868,qu'est il donc arrivé et pourquoi cet air de tristesse répandu sur tout votre bord
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0029.wav,49004,il est mort
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0040.wav,202828,après une longue conversation avec le commandant du port le capitaine leclère quitta naples fort agité

==> <target_dir>/ts_2018-09-28_fr_FR_test.csv <==
wav_filename,wav_filesize,transcript
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0001.wav,113836,chapitre un marseille l'arrivée
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0018.wav,104652,ah c'est vous dantès cria l'homme à la barque
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0028.wav,48172,il est mort
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0039.wav,123436,mon dieu monsieur de la façon la plus imprévue

==> <target_dir>/training_speech_2018-09-28_fr_FR_train.csv <==
wav_filename,wav_filesize,transcript
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0006.wav,105868,cependant ce bâtiment s'avançait
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0007.wav,282252,il avait heureusement franchi le détroit que quelque secousse volcanique a creusé entre l'île de calasareigne et l'île de jaros
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0009.wav,269708,néanmoins les experts en navigation reconnaissaient que si un accident était arrivé ce ne pouvait être au bâtiment lui même
<target_dir>/ts_2018-09-28_fr_FR/LeComteDeMonteCristoT1Chap1_0010.wav,156044,car il s'avançait dans toutes les conditions d'un navire parfaitement gouverné
...
```

Resulting alphabet is :

count | letter
------- | -----
 819470 | (space)
 551347 | e
 291556 | a
 278839 | s
 272008 | i
 258967 | t
 256397 | n
 253424 | r
 240781 | u
 218671 | o
 194406 | l
 134443 | d
 127131 | m
 116889 | c
  99572 | p
  75733 | v
  60234 | é
  53195 | '
  44553 | q
  40867 | f
  35339 | h
  35267 | b
  31629 | g
  27206 | j
  20068 | à
  14368 | x
  13043 | z
  12243 | è
   9974 | y
   8098 | ê
   3442 | ç
   2638 | â
   2464 | î
   2083 | ô
   1796 | û
   1494 | ù
   1208 | œ
     245 | ï
     156 | w
     136 | k
       45 | ü
       35 | ë
         2 | æ

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -19,3 +19,4 @@
 /native_client/python/utils_wrap.cpp
 /native_client/javascript/build
 /native_client/javascript/deepspeech_wrap.cxx
+.idea/
```
Comment (lissyx) [inline .gitignore]: Likely not wanted :)

**Comment by lissyx** [comment]
@kdavis-mozilla or @reuben want to give a look as well?

File: bin/import_training_speech.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    target_csv = os.path.join(target_dir, 'training_speech_' + ARCHIVE_NAME.replace('.zip', '.csv'))
+    if os.path.isfile(target_csv):
+        return
+    path_to_original_csv = os.path.join(extracted_dir, 'data.csv')
+    with open(path_to_original_csv) as csv_f:
+        data = [
+            dict(
+                wav_filename=os.path.join(target_dir, extracted_data, d['path']),
+                wav_filesize=os.path.getsize(os.path.join(target_dir, extracted_data, d['path'])),
+                transcript=d['text'],  # TODO: cleanup (ex: rm punctuation)
```
Comment (nicolaspanel) [inline bin/import_training_speech.py]: A cleanup is probably required here since transcript may contain punctuation (,.!? etc) and diatrics (éèêàç etc). Do you have an helper for that ?

**Comment by nicolaspanel** [comment]
@lissyx @kdavis-mozilla I now cleanup transcript to remove punctuations and capital letters.
Resulting alphabet is :

count | letter
------- | -----
 819470 | (space)
 551347 | e
 291556 | a
 278839 | s
 272008 | i
 258967 | t
 256397 | n
 253424 | r
 240781 | u
 218671 | o
 194406 | l
 134443 | d
 127131 | m
 116889 | c
  99572 | p
  75733 | v
  60234 | é
  53195 | '
  44553 | q
  40867 | f
  35339 | h
  35267 | b
  31629 | g
  27206 | j
  20068 | à
  14368 | x
  13043 | z
  12243 | è
   9974 | y
   8098 | ê
   3442 | ç
   2638 | â
   2464 | î
   2083 | ô
   1796 | û
   1494 | ù
   1208 | œ
    245 | ï
    156 | w
    136 | k
     45 | ü
     35 | ë
      2 | æ

is it ok for you ?

**Comment by lissyx** [comment]
WORKSFORME

**Comment by nicolaspanel** [comment]
@kdavis-mozilla fixed

File: bin/import_training_speech.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+    # Making path absolute
+    target_dir = path.abspath(target_dir)
+    # Conditionally download data
+    archive_path = _maybe_download(ARCHIVE_NAME, target_dir, ARCHIVE_URL)
+    # Conditionally extract common voice data
+    _maybe_extract(target_dir, ARCHIVE_DIR_NAME, archive_path)
+    # Conditionally convert common voice CSV files and mp3 data to DeepSpeech CSVs and wav
+    _maybe_convert_sets(target_dir, ARCHIVE_DIR_NAME)
+
+def _maybe_download(archive_name, target_dir, archive_url):
```
Comment (lissyx) [inline bin/import_training_speech.py]: @nicolaspanel Sorry, I just landed something to take care of that, you can remove that code :)

File: bin/import_training_speech.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
+#!/usr/bin/env python
```
Comment (lissyx) [inline bin/import_training_speech.py]: Looks like you might want to force python3 here

**Comment by nicolaspanel** [comment]
@lissyx should be ok now

**Comment by lissyx** [comment]
@nicolaspanel Looks like I can generate english-complatible `CSV` with just augmenting the `cleanup_transcription` with:
```
text = unicodedata.normalize('NFKD', text).encode('ASCII', 'ignore').decode()
```

File: .gitignore (hunk 1 above)
Comment (kdavis-mozilla) [reply .gitignore]: I'd guess this should be removed

File: bin/import_ts.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def _download_and_preprocess_data(target_dir):
+    # Making path absolute
+    target_dir = path.abspath(target_dir)
+    # Conditionally download data
+    archive_path = maybe_download('ts_' + ARCHIVE_NAME + '.zip', target_dir, ARCHIVE_URL)
+    # Conditionally extract common voice data
+    _maybe_extract(target_dir, ARCHIVE_DIR_NAME, archive_path)
+    # Conditionally convert common voice CSV files and mp3 data to DeepSpeech CSVs and wav
```
Comment (kdavis-mozilla) [inline bin/import_ts.py]: Forgot to change this comment;  it still refers to Common Voice

File: bin/import_ts.py [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+                dev_writer.writeheader()
+                test_writer = csv.DictWriter(test_csv_file, fieldnames=FIELDNAMES)
+                test_writer.writeheader()
+
+                for i, item in enumerate(data):
+                    transcript = validate_label(cleanup_transcript(item['text']))
+                    if not transcript:
+                        continue
+                    wav_filename = os.path.join(target_dir, extracted_data, item['path'])
+                    i_mod = i % 10
```
Comment (kdavis-mozilla) [inline bin/import_ts.py]: Nice trick, but is the order of the original file random?

If not this will introduce correlations in the train, dev, and test sets. For example if the original file was ordered by user, then a single user could land in train, dev, and test.

Comment (nicolaspanel) [reply bin/import_ts.py]: `data.csv` is ordered according to https://github.com/nicolaspanel/TrainingSpeech/blob/master/sources.json.

Since TrainingSpeech contains only a very few distinct speakers (4 right now, see https://github.com/nicolaspanel/TrainingSpeech#last-releases--download) the correlation will exists no matter how we split.

Comment (nicolaspanel) [reply bin/import_ts.py]: BTW, the purpose of https://github.com/nicolaspanel/TrainingSpeech is to provide additional training data (ie not dev / test data) since speeches are extracted from audiobooks and therefore may not represent modern/common way of speaking

Comment (nicolaspanel) [reply bin/import_ts.py]: @kdavis-mozilla 
I'm not saying we should drop the train/dev/test split. It is of course needed if your are using TrainingSpeech as your **only** datasource.

But, since TrainingSpeech samples are extracted from audiobooks and from a very few number of speakers, the performance on dev / test set will not represent real life performance such as noisy environments, familiar language, strong accents, etc.
A better approach would be to use dev / test sets that represent real life distributions. common-voice dev / test sets for example.

Therefore, IMOO, it is not the big deal if some (unwanted) correlations exists between train, dev et test.

However, I you prefer another approach  just tell me which one to use and I will send the changes
Best regards

Comment (kdavis-mozilla) [reply bin/import_ts.py]: I think with the size of this data set we can't get away from unwanted correlations. So we'll just have to live with them for now ¯\\_(ツ)_/¯

That said, when combining this data set with others to train on, e.g. this training set is added to a second training set, having a separate test set for this data set will allow us to get a feel for how the model trained on the total training set performs on this data set. (Of course due to the correlations mentioned above this will not be perfect.)

Comment (kdavis-mozilla) [reply bin/import_ts.py]: I should also say "real life performance" is use case specific. Someone may decide to use the model to transcribe audio books, then this data set reflects "real life performance" perfectly. So we should keep an open mind in this regard.

**Comment by lissyx** [comment]
Thanks @nicolaspanel that looks good! Can you squash your commits if some needs it and then rebase on master?

**Comment by nicolaspanel** [comment]
thanks @kdavis-mozilla 
@lissyx done
